### PR TITLE
Remove FutureMessage from RPC request callback logic

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50236 Apply clang-format to rpc cpp files
* #50029 Completely remove FutureMessage type
* #50028 Completely Remove FutureMessage from RPC agents
* #50027 Completely Remove FutureMessage from RPC cpp tests
* **#50026 Remove FutureMessage from RPC request callback logic**
* #50025 Completely remove FutureMessage from FaultyProcessGroupAgent
* #50024 Remove FutureMessage from sender TensorPipeAgent
* #50023 Remove FutureMessage from sender ProcessGroupAgent
* #50020 Completely remove FutureMessage from distributed autograd
* #50005 Completely remove FutureMessage from RPC TorchScript implementations
* #50004 Completely remove FutureMessage from RRef Implementations
* #49995 Replace FutureMessage with ivalue::Future in RpcAgent retry logic
* #49960 Replace FutureMessage with ivalue::Future in RRefContext
* #49927 Replace FutureMessage with ivalue::Future in distributed/autograd/utils.*
* #49906 Let RpcAgent::send() return JitFuture

Differential Revision: [D25753588](https://our.internmc.facebook.com/intern/diff/D25753588)

### CODE REVIEW

File: torch/csrc/distributed/rpc/request_callback_no_python.cpp [hunk 1]
```diff
@@ -398,22 +402,24 @@ void RequestCallbackNoPython::processForwardAutogradReq(
     TORCH_INTERNAL_ASSERT(wrappedRpcResponseFuture);
     if (wrappedRpcResponseFuture->hasError()) {
       // Propagate error to responseFuture if we had one.
-      responseFuture->setError(wrappedRpcResponseFuture->error()->what());
+      responseFuture->setError(wrappedRpcResponseFuture->exception_ptr());
     } else {
       auto msg = getMessageWithAutograd(
           fromWorkerId,
-          std::move(*wrappedRpcResponseFuture).moveValue(),
+          //std::move(*wrappedRpcResponseFuture).moveValue(),
```
Comment (lw) [inline torch/csrc/distributed/rpc/request_callback_no_python.cpp:409]: Is this comment a leftover?

File: torch/csrc/distributed/rpc/request_callback_no_python.cpp [hunk 2]
```diff
@@ -525,17 +533,18 @@ void RequestCallbackNoPython::processRunWithProfilingReq(
...
             // No need to propagate remote events in the case of an error.
-            responseFuture->setError(wrappedRpcResponseFuture->error()->what());
+            responseFuture->setError(wrappedRpcResponseFuture->exception_ptr());
           } else {
             populateRemoteProfiledEvents(
                 profiledEvents, profilingConfig, event_lists);
             auto rpcWithProfilingResp = std::make_unique<RpcWithProfilingResp>(
                 MessageType::RUN_WITH_PROFILING_RESP,
-                std::move(*wrappedRpcResponseFuture).moveValue(),
+                //std::move(*wrappedRpcResponseFuture).value(),
```
Comment (lw) [inline torch/csrc/distributed/rpc/request_callback_no_python.cpp:542]: Same here

File: torch/csrc/distributed/rpc/request_callback_no_python.cpp [hunk 3]
```diff
@@ -629,7 +639,7 @@ void RequestCallbackNoPython::processRpc(
   }
 }
 
-Message RequestCallbackNoPython::handleError(
+IValue RequestCallbackNoPython::handleError(
```
Comment (lw) [inline torch/csrc/distributed/rpc/request_callback_no_python.cpp:642]: What's the reason for changing this return type? It doesn't seem to be strictly necessary...

**Review by lw — APPROVED** [review]

File: torch/csrc/distributed/rpc/request_callback_no_python.cpp (hunk 3 above)
Comment (mrshenli) [reply torch/csrc/distributed/rpc/request_callback_no_python.cpp:642]: `handleError` was used in multiple places like this:

```
retFuture->markCompleted(handleError(e, request.type(), request.id()));
```

We can do the conversion in on the callsite, but doing it within `handleError` can help dedup some conversion code.

**Pull request closed without merging**